polyfill/lib/calendar.mjs

```diff
@@ -333,53 +332,1354 @@ function resolveNonLunisolarMonth(calendarDate) {
+        // the era or after its end as long as it's in the same year. If that
+        // happens, we'll adjust the era/eraYear pair to be the correct era for
+        // the `year`.
+        ({ eraYear, era } = eraFromYear(year));
+      } else {
+        throw new RangeError('Either `year` or `eraYear` and `era` are required');
+      }
+      return { ...calendarDate, year, eraYear, era };
+    },
+    adjustCalendarDate(calendarDate /*, cache, overflow, fromLegacyDate = false */) {
+      // Because this is not a lunisolar calendar, it's safe to convert monthCode to a number
+      const { month, monthCode } = calendarDate;
+      if (month === undefined) calendarDate = { ...calendarDate, month: monthCodeNumberPart(monthCode) };
+      this.validateCalendarDate(calendarDate);
+      calendarDate = this.completeEraYear(calendarDate);
+      calendarDate = nonIsoHelperBase.adjustCalendarDate(calendarDate);
+      return calendarDate;
+    },
+    estimateIsoDate(calendarDate) {
+      calendarDate = this.adjustCalendarDate(calendarDate);
+      const { year, month, day } = calendarDate;
+      const { anchorEra } = this;
+      const isoYearEstimate = year + anchorEra.isoEpoch.year - (anchorEra.hasYearZero ? 0 : 1);
+      return ES.RegulateDate(isoYearEstimate, month, day, 'constrain');
+    },
+    // Several calendars based on the Gregorian calendar use Julian dates (not
+    // proleptic Gregorian dates) before the Julian switchover in Oct 1582. See
+    // https://bugs.chromium.org/p/chromium/issues/detail?id=1173158.
+    checkJulianBug(calendarDate, isoDate) {
+      if (this.vulnerableToJulianBug) {
+        const beforeJulianSwitch = this.compareCalendarDates(isoDate, { year: 1582, month: 10, day: 15 }) < 0;
+        if (beforeJulianSwitch) {
+          const testDate = new Date('+001001-01-01T00:00Z').toLocaleDateString(`en-US-u-ca-${this.id}`, {
+            timeZone: 'UTC'
+          });
+          if (testDate.startsWith('12')) {
+            throw new RangeError(
+              `calendar '${this.id}' is broken for ISO dates before 1582-10-15` +
+                ' (see https://bugs.chromium.org/p/chromium/issues/detail?id=1173158)'
+            );
+          }
+        }
+      }
+    },
+    vulnerableToJulianBug: false
+  });
+};
+
+const makeHelperOrthodox = (id, originalEras) => {
+  const base = makeHelperGregorian(id, originalEras);
+  return ObjectAssign(base, {
+    inLeapYear(calendarDate /*, cache */) {
+      // Leap years happen one year before the Julian leap year. Note that this
+      // calendar is based on the Julian calendar which has a leap year every 4
+      // years, unlike the Gregorian calendar which doesn't have leap years on
+      // years divisible by 100 except years divisible by 400.
+      //
+      // Note that we're assuming that leap years in before-epoch times match
+      // how leap years are defined now. This is probably not accurate but I'm
+      // not sure how better to do it.
```

Comment:
Did you come across any prior art on how others approach this? I haven't found any yet; I suspect it's not well-defined, and/or the answer is "use the pre-reform calendar if you care about e.g. historical month/day accuracy before the epoch"

Reply:
To my knowledge, the approach is this way for the Julian calendar used before BCE by historians. And in fact the "historical" calendar does not match the Julian calendar, because from 45 BC until the beginning of the current era, the Roman pontifex put a leap year every 3 years instead of 4. Augustus corrected the error, that is why the month Sextilis was renamed in Augustus. But today everyone use and understands the Julian calendar as the calendar that follows the same rules before epoch.

Reply:
Nope. I also suspect that it's not well-defined. The overall problem is that lots of things (calendars, spelling, etc.) are very standardized now but weren't so much so 1000 or 2000 years ago.  If a medieval prince said that today was June 1, or that this year is a leap year, then I assume that was his prerogative. Time zones have a similar problem, but worse!  That said, it'd be helpful to specify our general philosophy for how to address pre-reform times.